Requires blueprint?

I just pulled 2.0.1 and I see that `MosaicZeroState` and `MosaicWindow` now has a dependency on @blueprintjs/core and @blueprintjs/icons. After I upgrade the version I get an error during build saying that it can't resolve @blueprintjs/icons and @blueprintjs/core. Is that intended? I saw PR #10 which removes the dependency...

### Bug report

- __Package version(s)__: 2.0.1
- __Browser and OS versions__: Windows, Chrome 

#### Steps to reproduce

1. I had version 1.1.1 of `react-mosaic-component` without any blueprintjs deps.
2. Changed `1.1.1` to `2.0.1` in package.json
3. `npm install`
4. `npm run build` which runs webpack

#### Actual behavior
```
ERROR in ./node_modules/react-mosaic-component/lib/MosaicWindow.js
Module not found: Error: Can't resolve '@blueprintjs/core' in '...\node_modules\react-mosaic-component\lib'
 @ ./node_modules/react-mosaic-component/lib/MosaicWindow.js 46:13-41
 @ ./node_modules/react-mosaic-component/lib/index.js

ERROR in ./node_modules/react-mosaic-component/lib/MosaicZeroState.js
Module not found: Error: Can't resolve '@blueprintjs/core' in '...\node_modules\react-mosaic-component\lib'
 @ ./node_modules/react-mosaic-component/lib/MosaicZeroState.js 35:13-41
 @ ./node_modules/react-mosaic-component/lib/index.js
```

#### Expected behavior

No build error, no need to install @blueprintjs/core and @blueprintjs/icons.

--- a/README.md
+++ b/README.md
@@ -33,7 +33,7 @@ making it easier for the consumer to style it to match their own app.
 By default, Mosaic renders with the `mosaic-blueprint-theme` class.
 This uses the excellent [Blueprint](http://blueprintjs.com/) React UI Toolkit to provide a good starting state.
 It is recommended to at least start developing with this theme.
-To use it install Blueprint `yarn add @blueprintjs/core` and add its CSS to your page.
+To use it install Blueprint `yarn add @blueprintjs/core @blueprintjs/icons` and add their CSS to your page.
 
 See [blueprint-theme.less](./styles/blueprint-theme.less) for an example of creating a theme.
 
@@ -61,7 +61,8 @@ body,
 
 ```tsx
 import { Mosaic } from 'react-mosaic-component';
-import '@blueprintjs/core/dist/blueprint.css';
+import '@blueprintjs/core/lib/css/blueprint.css';
+import '@blueprintjs/icons/lib/css/blueprint-icons.css';
 import './app.css';
 
 const ELEMENT_MAP: { [viewId: string]: JSX.Element } = {

--- a/README.md
+++ b/README.md
@@ -72,7 +72,7 @@ const ELEMENT_MAP: { [viewId: string]: JSX.Element } = {
 
 export const app = (
   <div id="app">
-    <Mosaic
+    <Mosaic<string>
       renderTile={(id) => ELEMENT_MAP[id]}
       initialValue={{
         direction: 'row',
@@ -105,11 +105,6 @@ The user can resize these panes but there is no other advanced functionality.
 ```tsx
 export type ViewId = 'a' | 'b' | 'c' | 'new';
 
-// Make type alias for generic checking in TSX until https://github.com/Microsoft/TypeScript/issues/6395 is fixed
-
-const ViewIdMosaic = Mosaic.ofType<ViewId>();
-const ViewIdMosaicWindow = MosaicWindow.ofType<ViewId>();
-
 const TITLE_MAP: Record<ViewId, string> = {
   a: 'Left Window',
   b: 'Top Right Window',
@@ -118,11 +113,11 @@ const TITLE_MAP: Record<ViewId, string> = {
 };
 
 export const app = (
-  <ViewIdMosaic
+  <Mosaic<ViewId>
     renderTile={(id, path) => (
-      <ViewIdMosaicWindow path={path} createNode={() => 'new'} title={TITLE_MAP[id]}>
+      <MosaicWindow<ViewId> path={path} createNode={() => 'new'} title={TITLE_MAP[id]}>
         <h1>{TITLE_MAP[id]}</h1>
-      </ViewIdMosaicWindow>
+      </MosaicWindow>
     )}
     initialValue={{
       direction: 'row',

--- a/circle.yml
+++ b/circle.yml
@@ -1,12 +1,11 @@
 machine:
   node:
-    version: 6.12.3
+    version: 10.13.0
 dependencies:
   pre:
     - npm install -g yarn
-    - yarn config set cache-folder /home/ubuntu/.cache/yarn
   cache_directories:
-    - "/home/ubuntu/.cache/yarn"
+    - "~/.cache/yarn"
     - "node_modules"
   override:
     - yarn install --frozen-lockfile

--- a/src/MosaicWindow.tsx
+++ b/src/MosaicWindow.tsx
@@ -14,8 +14,6 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-import { Classes, Icon } from '@blueprintjs/core';
-import { IconNames } from '@blueprintjs/icons';
 import classNames from 'classnames';
 import defer from 'lodash/defer';
 import dropRight from 'lodash/dropRight';
@@ -41,6 +39,7 @@ import { MosaicDropTarget } from './MosaicDropTarget';
 import { CreateNode, MosaicBranch, MosaicDirection, MosaicDragType, MosaicKey } from './types';
 import { createDragToUpdates } from './util/mosaicUpdates';
 import { getAndAssertNodeAtPathExists } from './util/mosaicUtilities';
+import { OptionalBlueprint } from './util/OptionalBlueprint';
 
 export interface MosaicWindowProps<T extends MosaicKey> {
   title: string;
@@ -95,7 +94,7 @@ export class InternalMosaicWindow<T extends MosaicKey> extends React.Component<
         </div>
         <div className="mosaic-window-body">
           <h4>{title}</h4>
-          <Icon iconSize={72} icon="application" />
+          <OptionalBlueprint.Icon iconSize={72} icon="application" />
         </div>
       </div>
     ),
@@ -195,13 +194,17 @@ export class InternalMosaicWindow<T extends MosaicKey> extends React.Component<
     return (
       <div className={classNames('mosaic-window-toolbar', { draggable: draggableAndNotRoot })}>
         {titleDiv}
-        <div className={classNames('mosaic-window-controls', Classes.BUTTON_GROUP)}>
+        <div className={classNames('mosaic-window-controls', OptionalBlueprint.getClasses('BUTTON_GROUP'))}>
           {hasAdditionalControls && (
             <button
               onClick={() => this.setAdditionalControlsOpen(!additionalControlsOpen)}
-              className={classNames(Classes.BUTTON, Classes.MINIMAL, Classes.iconClass(IconNames.MORE), {
-                [Classes.ACTIVE]: additionalControlsOpen,
-              })}
+              className={classNames(
+                OptionalBlueprint.getClasses('BUTTON', 'MINIMAL'),
+                OptionalBlueprint.getIconClass('MORE'),
+                {
+                  [OptionalBlueprint.getClasses('ACTIVE')]: additionalControlsOpen,
+                },
+              )}
             >
               <span className="control-text">{additionalControlButtonText!}</span>
             </button>

--- a/src/MosaicZeroState.tsx
+++ b/src/MosaicZeroState.tsx
@@ -14,14 +14,13 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-import { Classes, Icon } from '@blueprintjs/core';
-import { IconNames } from '@blueprintjs/icons';
 import classNames from 'classnames';
 import noop from 'lodash/noop';
 import React from 'react';
 
 import { MosaicActionsPropType, MosaicContext } from './contextTypes';
 import { CreateNode, MosaicKey } from './types';
+import { OptionalBlueprint } from './util/OptionalBlueprint';
 
 export interface MosaicZeroStateProps<T extends MosaicKey> {
   createNode?: CreateNode<T>;
@@ -36,14 +35,17 @@ export class MosaicZeroState<T extends MosaicKey> extends React.PureComponent<Mo
 
   render() {
     return (
-      <div className={classNames('mosaic-zero-state', Classes.NON_IDEAL_STATE)}>
-        <div className={Classes.NON_IDEAL_STATE_VISUAL}>
-          <Icon iconSize={120} icon="applications" />
+      <div className={classNames('mosaic-zero-state', OptionalBlueprint.getClasses('NON_IDEAL_STATE'))}>
+        <div className={OptionalBlueprint.getClasses('NON_IDEAL_STATE_VISUAL')}>
+          <OptionalBlueprint.Icon iconSize={120} icon="applications" />
         </div>
-        <h4 className={Classes.HEADING}>No Windows Present</h4>
+        <h4 className={OptionalBlueprint.getClasses('HEADING')}>No Windows Present</h4>
         <div>
           {this.props.createNode && (
-            <button className={classNames(Classes.BUTTON, Classes.iconClass(IconNames.ADD))} onClick={this.replace}>
+            <button
+              className={classNames(OptionalBlueprint.getClasses('BUTTON'), OptionalBlueprint.getIconClass('ADD'))}
+              onClick={this.replace}
+            >
               Add New Window
             </button>
           )}

--- a/src/buttons/ExpandButton.tsx
+++ b/src/buttons/ExpandButton.tsx
@@ -14,12 +14,12 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-import { Classes } from '@blueprintjs/core';
-import { IconNames } from '@blueprintjs/icons';
+import classNames from 'classnames';
 import React from 'react';
 
 import { MosaicWindowContext } from '../contextTypes';
 import { MosaicKey } from '../types';
+import { OptionalBlueprint } from '../util/OptionalBlueprint';
 import { createDefaultToolbarButton, MosaicButtonProps } from './MosaicButton';
 
 export class ExpandButton<T extends MosaicKey> extends React.PureComponent<MosaicButtonProps> {
@@ -27,7 +27,11 @@ export class ExpandButton<T extends MosaicKey> extends React.PureComponent<Mosai
   context!: MosaicWindowContext<T>;
 
   render() {
-    return createDefaultToolbarButton('Expand', Classes.iconClass(IconNames.MAXIMIZE), this.expand);
+    return createDefaultToolbarButton(
+      'Expand',
+      classNames('expand-button', OptionalBlueprint.getIconClass('MAXIMIZE')),
+      this.expand,
+    );
   }
 
   private expand = () => {

--- a/src/buttons/MosaicButton.tsx
+++ b/src/buttons/MosaicButton.tsx
@@ -14,9 +14,10 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-import { Classes } from '@blueprintjs/core';
 import classNames from 'classnames';
 import React from 'react';
+
+import { OptionalBlueprint } from '../util/OptionalBlueprint';
 
 export function createDefaultToolbarButton(
   title: string,
@@ -28,9 +29,9 @@ export function createDefaultToolbarButton(
     <button
       title={title}
       onClick={onClick}
-      className={classNames('mosaic-default-control', Classes.BUTTON, Classes.MINIMAL, className)}
+      className={classNames('mosaic-default-control', OptionalBlueprint.getClasses('BUTTON', 'MINIMAL'), className)}
     >
-      {text! && <span className="control-text">text!</span>}
+      {text && <span className="control-text">{text}</span>}
     </button>
   );
 }

--- a/src/buttons/RemoveButton.tsx
+++ b/src/buttons/RemoveButton.tsx
@@ -14,12 +14,12 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-import { Classes } from '@blueprintjs/core';
-import { IconNames } from '@blueprintjs/icons';
+import classNames from 'classnames';
 import React from 'react';
 
 import { MosaicWindowContext } from '../contextTypes';
 import { MosaicKey } from '../types';
+import { OptionalBlueprint } from '../util/OptionalBlueprint';
 import { createDefaultToolbarButton, MosaicButtonProps } from './MosaicButton';
 
 export class RemoveButton<T extends MosaicKey> extends React.PureComponent<MosaicButtonProps> {
@@ -27,7 +27,11 @@ export class RemoveButton<T extends MosaicKey> extends React.PureComponent<Mosai
   context!: MosaicWindowContext<T>;
 
   render() {
-    return createDefaultToolbarButton('Close Window', Classes.iconClass(IconNames.CROSS), this.remove);
+    return createDefaultToolbarButton(
+      'Close Window',
+      classNames('close-button', OptionalBlueprint.getIconClass('CROSS')),
+      this.remove,
+    );
   }
 
   private remove = () => {

--- a/src/buttons/ReplaceButton.tsx
+++ b/src/buttons/ReplaceButton.tsx
@@ -14,13 +14,13 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-import { Classes } from '@blueprintjs/core';
-import { IconNames } from '@blueprintjs/icons';
+import classNames from 'classnames';
 import noop from 'lodash/noop';
 import React from 'react';
 
 import { MosaicWindowContext } from '../contextTypes';
 import { MosaicKey } from '../types';
+import { OptionalBlueprint } from '../util/OptionalBlueprint';
 import { createDefaultToolbarButton, MosaicButtonProps } from './MosaicButton';
 
 export class ReplaceButton<T extends MosaicKey> extends React.PureComponent<MosaicButtonProps> {
@@ -28,7 +28,11 @@ export class ReplaceButton<T extends MosaicKey> extends React.PureComponent<Mosa
   context!: MosaicWindowContext<T>;
 
   render() {
-    return createDefaultToolbarButton('Replace Window', Classes.iconClass(IconNames.EXCHANGE), this.replace);
+    return createDefaultToolbarButton(
+      'Replace Window',
+      classNames('replace-button', OptionalBlueprint.getIconClass('EXCHANGE')),
+      this.replace,
+    );
   }
 
   private replace = () => {

--- a/src/buttons/SplitButton.tsx
+++ b/src/buttons/SplitButton.tsx
@@ -14,13 +14,13 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-import { Classes } from '@blueprintjs/core';
-import { IconNames } from '@blueprintjs/icons';
+import classNames from 'classnames';
 import noop from 'lodash/noop';
 import React from 'react';
 
 import { MosaicWindowContext } from '../contextTypes';
 import { MosaicKey } from '../types';
+import { OptionalBlueprint } from '../util/OptionalBlueprint';
 import { createDefaultToolbarButton, MosaicButtonProps } from './MosaicButton';
 
 export class SplitButton<T extends MosaicKey> extends React.PureComponent<MosaicButtonProps> {
@@ -28,7 +28,11 @@ export class SplitButton<T extends MosaicKey> extends React.PureComponent<Mosaic
   context!: MosaicWindowContext<T>;
 
   render() {
-    return createDefaultToolbarButton('Split Window', Classes.iconClass(IconNames.ADD_COLUMN_RIGHT), this.split);
+    return createDefaultToolbarButton(
+      'Split Window',
+      classNames('split-button', OptionalBlueprint.getIconClass('ADD_COLUMN_RIGHT')),
+      this.split,
+    );
   }
 
   private split = () => {

--- a/src/util/OptionalBlueprint.tsx
+++ b/src/util/OptionalBlueprint.tsx
@@ -1,0 +1,34 @@
+import { Classes as ClassesImport, IIconProps } from '@blueprintjs/core';
+import { IconNames as IconNamesTypeImport } from '@blueprintjs/icons';
+import * as React from 'react';
+
+export namespace OptionalBlueprint {
+  type Classes = typeof ClassesImport;
+  let Classes: Classes | undefined;
+  export let Icon: React.ReactType<IIconProps>;
+  type IconNames = typeof IconNamesTypeImport;
+  let IconNames: IconNames | undefined;
+  try {
+    ({ Classes, Icon } = require('@blueprintjs/core'));
+    ({ IconNames } = require('@blueprintjs/icons'));
+  } catch {
+    Icon = ({ icon }: IIconProps) => <span>{icon}</span>;
+  }
+
+  type BlueprintClass = { [K in keyof Classes]: Classes[K] extends string ? K : never }[keyof Classes];
+  export function getClasses(...names: BlueprintClass[]): string {
+    if (Classes) {
+      return names.map((name) => Classes![name]).join(' ');
+    }
+
+    return '';
+  }
+
+  export function getIconClass(iconName: keyof IconNames): string {
+    if (Classes && IconNames) {
+      return Classes.iconClass(IconNames[iconName]);
+    }
+
+    return '';
+  }
+}

--- a/styles/mosaic-window.less
+++ b/styles/mosaic-window.less
@@ -128,16 +128,16 @@
 
 .mosaic:not(.mosaic-blueprint-theme) {
   .mosaic-default-control {
-    &.@{ns}-icon-cross:before {
+    &.close-button:before {
       content: 'Close';
     }
-    &.@{ns}-icon-add-column-right:before {
+    &.split-button:before {
       content: 'Split';
     }
-    &.@{ns}-icon-exchange:before {
+    &.replace-button:before {
       content: 'Replace';
     }
-    &.@{ns}-icon-maximize:before {
+    &.expand-button:before {
       content: 'Expand';
     }
   }

